relayer/workers/ethrelayer/parachain-writer.go

```diff
@@ -26,51 +30,53 @@ type ParachainWriter struct {
 
 func NewParachainWriter(
 	conn *parachain.Connection,
-	messages <-chan []chain.Message,
-	headers <-chan chain.Header,
+	payloads <-chan ParachainPayload,
 	log *logrus.Entry,
 ) *ParachainWriter {
 	return &ParachainWriter{
 		conn:     conn,
-		messages: messages,
-		headers:  headers,
+		payloads: payloads,
 		log:      log,
 	}
 }
 
 func (wr *ParachainWriter) Start(ctx context.Context, eg *errgroup.Group) error {
+	cancelWithError := func(err error) error {
+		// Ensures the context is canceled so that the channels below are
+		// closed by the listener
+		eg.Go(func() error { return err })
+
+		wr.log.Info("Shutting down writer...")
+		// Avoid deadlock if the listener is still trying to send to a channel
+		for range wr.payloads {
```

Comment:
At a high-level, this does have a neat effect in causing an orderly teardown of the producer/consumer loop. Since we've previously cancelled the errgoup/context, This effectively blocks until the producer gets around to closing the channel and killing itself. :+1:

Reply:
I think this pattern is really key to help us avoid deadlocks in future